storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -3108,10 +3108,1655 @@ buf_block_t *fsp_get_latched_xdes_page(
+      if (state >> 31)
+        goto func_exit;
+
+      switch (state) {
+      case XDES_FREE:
+        goto prev_extent;
+      case XDES_FSEG:
+      case XDES_FULL_FRAG:
+      case XDES_FREE_FRAG:
+        uint32_t dest= find_free_extent(free_limit);
+	if (dest == FIL_NULL)
+	  goto func_exit;
+	m_extent_map[free_limit]= dest;
+	break;
+      }
+prev_extent:
+      free_limit-= FSP_EXTENT_SIZE;
+    }
+func_exit:
+    if (m_extent_map.empty())
+      return false;
+
+    sql_print_information("InnoDB: System tablespace defragmentation "
+                          "process starts");
+    for (const auto &kv : m_extent_map)
+      sql_print_information("InnoDB: Moving the data from old "
+		            "extent %d to %d", kv.first, kv.second);
+    return true;
+  }
+
+  /** Defragment the system tables */
+  dberr_t defragment_system_tables() noexcept
+  {
+    dberr_t err= defragment_table(dict_sys.sys_tables);
+    if (err == DB_SUCCESS)
+      err= defragment_table(dict_sys.sys_columns);
+    if (err == DB_SUCCESS)
+      err= defragment_table(dict_sys.sys_indexes);
+    if (err == DB_SUCCESS)
+      err= defragment_table(dict_sys.sys_fields);
+    if (err == DB_SUCCESS)
+      err= defragment_table(dict_sys.sys_foreign);
+    if (err == DB_SUCCESS)
+      err= defragment_table(dict_sys.sys_foreign_cols);
+    if (err == DB_SUCCESS)
+      err= defragment_table(dict_sys.sys_virtual);
+    return err;
+  }
+
+  /** @return extent which replaces the later extent
+  or same extent if there is no replacement exist */
+  uint32_t get_new_extent(uint32_t old_extent) const noexcept
+  {
+    auto it= m_extent_map.find(old_extent);
+    if (it != m_extent_map.end())
+      return it->second;
+    return old_extent;
+  }
+
+  /** @return state for the given extent */
+  uint32_t get_state(uint32_t extent) noexcept
+  {
+    return m_extent_info[extent];
+  }
+};
+
+#ifdef UNIV_DEBUG
+/** Validate the system tablespace list */
+__attribute__((warn_unused_result))
+dberr_t fsp_tablespace_validate(fil_space_t *space)
+{
+  /* Validate all FSP list in system tablespace */
+  mtr_t local_mtr;
+  dberr_t err= DB_SUCCESS;
+  local_mtr.start();
+  if (buf_block_t *header= fsp_get_header(
+        space, &local_mtr, &err))
+  {
+    flst_validate(header, FSP_FREE + FSP_HEADER_OFFSET, &local_mtr);
+    flst_validate(header, FSP_FREE_FRAG + FSP_HEADER_OFFSET,
+                  &local_mtr);
+    flst_validate(header, FSP_HEADER_OFFSET + FSP_FULL_FRAG,
+                  &local_mtr);
+    flst_validate(header, FSP_HEADER_OFFSET + FSP_SEG_INODES_FULL,
+                  &local_mtr);
+    flst_validate(header, FSP_HEADER_OFFSET + FSP_SEG_INODES_FREE,
+                  &local_mtr);
+  }
+  local_mtr.commit();
+  return err;
+}
+
+/** Validate the system tablespace list */
+__attribute__((warn_unused_result))
+dberr_t fseg_validate(fil_space_t *space, dict_index_t *index)
+{
+  /* Validate all FSP list in system tablespace */
+  mtr_t local_mtr;
```

Comment:
This function definition is missing `static`. It is not being invoked from outside this compilation unit.

When there is no function parameter named  `mtr`, it would be simpler to use the local variable name `mtr` instead of `local_mtr`.